src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,263 @@
+
+/*
+ * Setting that controls whether replication origin sessions are enabled.
+ */
+bool UseReplicationOriginsForInternalTransfers = false;
+
+
+/* replication_origin_session_start_no_publish starts a new replication origin session
+ * in the local node. This function is used to avoid publishing the WAL records to the
+ * replication slot by setting replication origin to DoNotReplicateId in WAL records.
+ * It remembers the previous replication origin for the current session which will be
+ * used to reset the replication origin to the previous value when the session ends.
+ */
+Datum
+replication_origin_session_start_no_publish(PG_FUNCTION_ARGS)
+{
+	if (!UseReplicationOriginsForInternalTransfers)
+	{
+		PG_RETURN_VOID();
+	}
+	SetupReplicationOriginLocalSession();
+	PG_RETURN_VOID();
+}
+
+
+/* replication_origin_session_end_no_publish ends the current replication origin session
+ * in the local node. This function is used to reset the replication origin to the
+ * earlier value of replication origin.
+ */
+Datum
+replication_origin_session_end_no_publish(PG_FUNCTION_ARGS)
+{
+	ResetReplicationOriginLocalSession();
+	PG_RETURN_VOID();
+}
+
+
+/* replication_origin_session_is_no_publish checks if the current replication origin
+ * session is active in the local node.
+ */
+Datum
+replication_origin_session_is_no_publish(PG_FUNCTION_ARGS)
+{
+	bool result = IsLocalReplicationOriginSessionActive();
+	PG_RETURN_BOOL(result);
+}
+
+
+/* IsLocalReplicationOriginSessionActive checks if the current replication origin
+ * session is active in the local node.
+ */
+inline bool
+IsLocalReplicationOriginSessionActive(void)
+{
+	return (replorigin_session_origin != InvalidRepOriginId);
+}
+
+
+/*
+ * SetupReplicationOriginLocalSession sets up a new replication origin session in a
+ * local session.
+ */
+void
+SetupReplicationOriginLocalSession(void)
+{
+	if (!UseReplicationOriginsForInternalTransfers)
+	{
+		return;
+	}
+
+	/*elog(LOG, "Setting up local replication origin session"); */
+	if (!IsLocalReplicationOriginSessionActive())
+	{
+		originalOriginId = replorigin_session_origin;
+		replorigin_session_origin = DoNotReplicateId;
+
+		/* Register a call back for ResetReplicationOriginLocalSession function for error cases */
+		MemoryContextCallback *replicationOriginResetCallback = palloc0(
+			sizeof(MemoryContextCallback));
+		replicationOriginResetCallback->func =
+			ResetReplicationOriginLocalSessionCallbackHandler;
+		replicationOriginResetCallback->arg = NULL;
+		MemoryContextRegisterResetCallback(CurrentMemoryContext,
+										   replicationOriginResetCallback);
+	}
+}
+
+
+/*
+ * ResetReplicationOriginLocalSession resets the replication origin session in a
+ * local node.
+ */
+void
+ResetReplicationOriginLocalSession(void)
+{
+	/*elog(LOG, "Resetting local replication origin session"); */
+	if (replorigin_session_origin != DoNotReplicateId)
+	{
+		return;
+	}
+
+	if (IsLocalReplicationOriginSessionActive())
```

Comment:
seems redundant, since we only get here if replorigin_session_origin == DoNotReplicateId

Reply:
OK removed it